dll/win32/iernonce/registry.cpp

```diff
@@ -0,0 +1,186 @@
+}
+
+RunOnceExEntry::RunOnceExEntry(const ATL::CStringW &Name, const ATL::CStringW &Value) :
+    m_Name(Name), m_Value(Value)
+{ ; }
+
+int RunOnceExEntryCmp(const void *a, const void *b)
+{
+    return lstrcmpW(((RunOnceExEntry *)a)->m_Name,
+        ((RunOnceExEntry *)b)->m_Name);
+}
+
+BOOL RunOnceExSection::HandleValue(CRegKeyEx &hKey, const CStringW &ValueName, DWORD cchszName)
+{
+    DWORD dwType;
+    DWORD cbData;
+    // Query data size
+    if (hKey.QueryValue(ValueName, &dwType, NULL, &cbData) != ERROR_SUCCESS)
+        return FALSE;
+
+    // Validate its format and size.
+    if (dwType != REG_SZ)
+        return TRUE;
+
+    if (cbData % sizeof(WCHAR) != 0)
+        return FALSE;
+
+    CStringW Buffer;
+    LPWSTR szBuffer = Buffer.GetBuffer((cbData / sizeof(WCHAR)) + 1);
+
+    if (hKey.QueryValue(ValueName, &dwType, szBuffer, &cbData) != ERROR_SUCCESS)
+    {
+        Buffer.ReleaseBuffer();
+        return FALSE;
+    }
+    szBuffer[cbData / sizeof(WCHAR)] = L'\0';
+    Buffer.ReleaseBuffer();
+
+    if (ValueName.IsEmpty())
+    {
+        // this is the default value
+        m_SectionTitle = Buffer;
+    }
+    else
+    {
+        m_EntryList.Add(RunOnceExEntry(ValueName, Buffer));
+    }
+
+    return TRUE;
+}
+
+RunOnceExSection::RunOnceExSection(CRegKeyEx &hParentKey, const CStringW &lpSubKeyName) : m_SectionName(lpSubKeyName)
+{
+    m_bSuccess = FALSE;
+    CRegKeyEx hKey;
+
+    if (hKey.Open(hParentKey, lpSubKeyName, KEY_READ) != ERROR_SUCCESS)
+        return;
+
+    DWORD dwValueNum;
+    DWORD dwMaxValueNameLen;
+    LSTATUS Error;
+
+    Error = RegQueryInfoKeyW(hKey, NULL, 0, NULL,
+        NULL, NULL, NULL, &dwValueNum, &dwMaxValueNameLen, NULL, NULL, NULL);
+
+    if (Error != ERROR_SUCCESS)
+        return;
+
+    CStringW ValueName;
+    LPWSTR szValueName;
+
+    for (DWORD i = 0; i < dwValueNum; i++)
+    {
+        DWORD dwcchName = dwMaxValueNameLen + 1;
+
+        szValueName = ValueName.GetBuffer(dwMaxValueNameLen + 1);
+        Error = hKey.EnumValueName(i, szValueName, &dwcchName);
+        ValueName.ReleaseBuffer();
+
+        if (Error != ERROR_SUCCESS)
+        {
+            // TODO: error handling
+            return;
+        }
+
+        if (!HandleValue(hKey, ValueName, dwcchName + 1))
+            return;
+    }
+
+    // sort entries by name in string order.
+    qsort(m_EntryList.GetData(), m_EntryList.GetSize(), sizeof(RunOnceExEntry),
+          RunOnceExEntryCmp);
+
+    hKey.Close();
+    m_bSuccess = TRUE;
+    return;
+}
+
+int RunOnceExSectionCmp(const void *a, const void *b)
+{
+    return lstrcmpW(((RunOnceExSection *)a)->m_SectionName,
+        ((RunOnceExSection *)b)->m_SectionName);
+}
+
+RunOnceExInstance::RunOnceExInstance(HKEY BaseKey)
+{
+    m_bSuccess = FALSE;
+    CRegKeyEx hKey;
+
+    if (hKey.Open(BaseKey,
+        L"Software\\Microsoft\\Windows\\CurrentVersion\\RunOnceEx\\", KEY_READ)
+        != ERROR_SUCCESS)
+        return;
```

Comment:
Here and in other places: See https://reactos.org/wiki/Coding_Style#Indentation_and_line_width point 4 and others.

Reply:
should I align it like this:

```c
if (hKey.Open(BaseKey,
              L"Software\\Microsoft\\Windows\\CurrentVersion\\RunOnceEx\\",
              KEY_READ) != ERROR_SUCCESS)
```
🤣

Reply:
yes, that's the spirit (and the other places)

Reply:
ok! I will pay attention!